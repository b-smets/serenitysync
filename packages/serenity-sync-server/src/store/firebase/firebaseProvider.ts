import * as firebase from 'firebase-admin';

const serviceAccount = require('../../../config/firebase.json');

const app = firebase.initializeApp({
  credential: firebase.credential.cert(serviceAccount),
  databaseURL: `https://${serviceAccount.project_id}.firebaseio.com`,
});

export const documentsRootRef = (databaseName: string) =>
  app.database().ref(`data/${databaseName}/documents`);
export const documentRef = (databaseName: string, id: string) =>
  app.database().ref(`data/${databaseName}/documents/${id}`);
