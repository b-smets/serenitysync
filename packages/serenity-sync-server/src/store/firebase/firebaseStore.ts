import * as firebase from 'firebase-admin';
import { DocumentRevision } from '../../document/document';
import { Store } from '../store';
import { documentRef, documentsRootRef } from './firebaseProvider';

export class FirebaseStore implements Store {

  allDocuments(databaseName: string): Promise<DocumentRevision[]> {
    return documentsRootRef(databaseName).once('value')
      .then((snap: firebase.database.DataSnapshot) => {
        if (snap.exists()) {
          const value = snap.val();
          return Object.keys(value).map(key => value[key]);
        }
        return [];
      });
  }

  document(databaseName: string, id: string): Promise<DocumentRevision | undefined> {
    return documentRef(databaseName, id).once('value')
      .then((snap: firebase.database.DataSnapshot) => {
        if (snap.exists()) {
          return snap.val();
        }
        return undefined;
      });
  }

  setDocument(databaseName: string, document: DocumentRevision): Promise<void> {
    return documentRef(databaseName, document.id).set(document);
  }
}
