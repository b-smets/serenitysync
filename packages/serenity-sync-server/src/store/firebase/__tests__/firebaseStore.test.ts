jest.unmock('../firebaseStore');
jest.unmock('firebase-mock');
jest.mock('../firebaseProvider', () => ({
  documentsRootRef: jest.fn(),
  documentRef: jest.fn(),
}));

import { MockFirebase } from 'firebase-mock';
import { documentRef, documentsRootRef } from '../firebaseProvider';
import { FirebaseStore } from '../firebaseStore';

describe('FirebaseStore', () => {

  const store = new FirebaseStore();
  let database: MockFirebase;

  const initializeDatabase = (data: any) => {
    database = new MockFirebase('https://firebaseio.com', data).autoFlush(0);
  };

  beforeAll(() => {
    (documentsRootRef as jest.Mock<any>).mockImplementation(() => database);
    (documentRef as jest.Mock<any>).mockImplementation(() => database);
  });

  beforeEach(() => {
    initializeDatabase({});
  });

  describe('allDocuments', () => {

    test('uses the correct database document root ref', () => {
      return store.allDocuments('mydb').then(result => {
        expect(documentsRootRef).toBeCalledWith('mydb');
      });
    });

    test('returns a list of documents', () => {
      initializeDatabase({
        foo: {
          id: 'foo',
          rev: '1-abc',
          body: {
            one: 'alpha',
            two: 'beta',
            three: 'gamma',
          },
        },
        bar: {
          id: 'bar',
          rev: '2-def',
          body: {
            three: 'delta',
            four: 'epsilon',
            five: 'omega',
          },
        },
      });
      return store.allDocuments('mydb').then(result => {
        expect(result).toMatchSnapshot();
      });
    });

    test('returns an empty list if the database contains no documents', () => {
      return store.allDocuments('mydb').then(result => {
        expect(result).toEqual([]);
      });
    });

    test('returns an empty list if the database does not exist', () => {
      initializeDatabase(undefined);
      return store.allDocuments('mydb').then(result => {
        expect(result).toEqual([]);
      });
    });
  });

  describe('document', () => {

    test('uses the correct document ref', () => {
      return store.document('mydb', 'foo').then(result => {
        expect(documentRef).toBeCalledWith('mydb', 'foo');
      });
    });

    test('returns the document if it exists', () => {
      const document = {
        id: 'foo',
        rev: '1-abc',
        body: {
          one: 'alpha',
          two: 'beta',
          three: 'gamma',
        },
      };

      initializeDatabase(document);
      return store.document('mydb', 'foo').then(result => {
        expect(result).toEqual(document);
      });
    });

    test('returns undefined if the document does not exist', () => {
      return store.document('mydb', 'foo').then(result => {
        expect(result).toBeUndefined();
      });
    });
  });

  describe('setDocument', () => {

    const document = {
      id: 'foo',
      rev: '1-abc',
      body: {
        one: 'alpha',
        two: 'beta',
        three: 'gamma',
      },
    };

    test('uses the correct document ref', () => {
      return store.setDocument('mydb', document).then(result => {
        expect(documentRef).toBeCalledWith('mydb', 'foo');
      });
    });

    test('sets the document if it did not yet exist', () => {
      const spy = jest.spyOn(database, 'set');
      return store.setDocument('mydb', document).then(result => {
        expect(spy).toBeCalledWith(document);
      });
    });
  });
});
