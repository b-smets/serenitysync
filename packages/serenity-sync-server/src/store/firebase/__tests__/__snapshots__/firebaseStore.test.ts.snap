// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FirebaseStore allDocuments returns a list of documents 1`] = `
Array [
  Object {
    "body": Object {
      "five": "omega",
      "four": "epsilon",
      "three": "delta",
    },
    "id": "bar",
    "rev": "2-def",
  },
  Object {
    "body": Object {
      "one": "alpha",
      "three": "gamma",
      "two": "beta",
    },
    "id": "foo",
    "rev": "1-abc",
  },
]
`;
